torchtune/models/llama4/_component_builders.py

```diff
@@ -0,0 +1,349 @@
+# Copyright (c) Meta Platforms, Inc. and affiliates.
+# All rights reserved.
+#
+# This source code is licensed under the BSD-style license found in the
+# LICENSE file in the root directory of this source tree.
+
+from functools import partial
+from typing import List, Optional
+
+from torch import nn
+from torchtune.models.clip._component_builders import clip_vision_encoder
+from torchtune.models.llama4._chunked_attention import get_chunked_attention_mask
+
+from torchtune.models.llama4._encoder import (
+    Llama4VisionEncoder,
+    Llama4VisionProjectionHead,
+)
+from torchtune.models.llama4._position_embeddings import Llama4ScaledRoPE
+from torchtune.modules import (
+    FeedForward,
+    FrozenNF4Linear,
+    MultiHeadAttention,
+    rms_norm,
+    RMSNorm,
+    RotaryPositionalEmbeddings,
+    TransformerDecoder,
+    TransformerSelfAttentionLayer,
+)
+from torchtune.modules.moe import GroupedExperts, MoE, TokenChoiceTopKRouter
+
+
+"""
+Component builders for the Llama4 model.
+
+torchtune provides composable building blocks. Builder functions help
+stitch these building blocks into higher-level components. This design has
+two benefits:
+- The building blocks themselves are very flexible. For example, ``MultiHeadAttention``
+can take either nn.Linear or nn.LoRALinear for ``q_proj``.
+- Builder functions expose a set of configurable params which keep the constructors of
+the building blocks simple.
+"""
+
+
+def llama4_vision_encoder(
+    # clip encoder parameters
+    *,
+    patch_size: int,
+    num_heads: int,
+    clip_embed_dim: int,
+    clip_num_layers: int,
+    clip_hidden_states: Optional[List[int]] = None,
+    # projection parameters
+    projection_embed_dim: int,
+    decoder_embed_dim: int,
+    # image parameters
+    tile_size: int,
+    max_num_tiles: int = 1,
+    in_channels: int = 3,
+) -> Llama4VisionEncoder:
+    """
+    Build the Llama 4 vision encoder by combining the CLIP image model with an additional
+    projection head fusion module. This includes:
+    - Spatial positional encodings
+    - CLIP model backbone
+    - Projection head on top of CLIP
+    - Final projection into token embedding dimension
```

Comment:
this is no longer true right? it's just the MLP projection head?

Reply:
Ah yeah will update